### PR TITLE
Add new case vioinput_mice.device_mouse/device_tablet

### PR DESCRIPTION
**Pull request by peixiu**
Two new case:
vioinput_mice.device_mouse
vioinput_mice.device_table

id: 1724060, 1724061
Signed-off-by: Peixiu Hou <phou@redhat.com>

### CODE REVIEW

**Comment by peixiu** [comment]
(1/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_mouse_basic.device_mouse.q35: PASS (415.34 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_mouse_basic.device_tablet.q35: PASS (296.04 s)

**Comment by peixiu** [comment]
(1/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_mouse_basic.device_mouse.q35: PASS (414.93 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_mouse_basic.device_tablet.q35: PASS (244.18 s)

**Comment by peixiu** [comment]
(1/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_mouse_basic.device_mouse.q35: PASS (466.10 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.input_mouse_basic.device_tablet.q35: PASS (243.72 s)

@luckyh When you free, please help to review this patch~

Thanks a lot~

**Comment by vivianQizhu** [comment]
@peixiu Please fix CI failure.

**Comment by peixiu** [comment]
(1/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (92.93 s)
 (2/2) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (93.82 s)

**Comment by peixiu** [comment]
@luckyh I updated the code as bug172406 confirmed solution, both device_tablet and device_mouse used same flow, please help to review it when you free, thanks~

File: qemu/tests/vioinput_mice.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    driver_name = params["driver_name"]
+    extra_driver_verify = params.objects("extra_driver_verify")
+
+    if params["os_type"] == "windows":
+        session = vm.wait_for_login()
+
+        error_context.context("Check vioinput driver is running", logging.info)
+        utils_test.qemu. windrv_verify_running(session, test, driver_name)
```
[the remark this answers is not in the document]
Comment (peixiu) [reply qemu/tests/vioinput_mice.py]: my mistake, will update it, thanks~

File: qemu/tests/vioinput_mice.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+    mouse_btn_test(test, params, console, listener, wait_time)
+    mouse_scroll_test(test, params, console, listener, wait_time)
+
+    if not params.get("target_pos", None):
+        width, height = console.screen_size()
+        x_max, y_max = width-1, height-1
+        target_pos = [(0, 0), (x_max, 0), (0, y_max), (x_max, y_max)]
+    else:
+        target_pos = params.get("target_pos")
```
[the remark this answers is not in the document]
Comment (peixiu) [reply qemu/tests/vioinput_mice.py]: my mistake, will update it, thanks~

File: qemu/tests/vioinput_mice.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    width, height = console.screen_size()
+    events_queue = listener.events
+    event_lst = []
+    start_pos = console.pointer_position()
+    x0, y0 = start_pos
+    xn, yn = end_pos
+    # Compute a line y=kx+b through start_pos and end_pos.
+    if (xn - x0) != 0:
+        vertical = 0
+        k = (yn - y0) / (xn - x0)
```
[the remark this answers is not in the document]
Comment (peixiu) [reply qemu/tests/vioinput_mice.py]: yes, thanks for point, will update it~

File: qemu/tests/vioinput_mice.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        b = yn - (k * xn)
+    else:
+        vertical = 1
+
+    error_context.context("Moving pointer from %s to %s" % (start_pos, end_pos))
+    console.pointer_move(end_pos, motion=line, mtype=mtype)
+    time.sleep(wait_time)
```
[the remark this answers is not in the document]
Comment (peixiu) [reply qemu/tests/vioinput_mice.py]: ok, will update it, thanks

File: qemu/tests/vioinput_mice.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+        # Filter beyond screen size events.
+        # Due to os will ignores/corrects these events.
+        if 0 <= xpos <= width and 0 <= ypos <= height:
+            event_lst.append((event["xPos"], event["yPos"]))
+
+    xn_guest, yn_guest = event_lst[-1]
+    tolerance = int(params.get("tolerance", 5))
+    error_context.context("Compare if pointer move to destination pos (%s, %s)"
+                          "the missed value should in tolerance scope." % end_pos)
+    if (abs(xn - xn_guest) > tolerance) or (abs(yn - yn_guest) > tolerance):
```
[the remark this answers is not in the document]
Comment (peixiu) [reply qemu/tests/vioinput_mice.py]: for rel device, it's not  necessary, but better check this as well。
due to it has much missed, so I set a large tolerance value for it according to test experience.
I can also do more tests for it, if the tolerance need too large, thus this test will meaningless, will cancel it, thanks~

File: qemu/tests/vioinput_mice.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (peixiu) [reply qemu/tests/vioinput_mice.py]: updated~

File: qemu/tests/vioinput_mice.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (peixiu) [reply qemu/tests/vioinput_mice.py]: updated~

File: qemu/tests/vioinput_mice.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (peixiu) [reply qemu/tests/vioinput_mice.py]: updated~

File: qemu/tests/vioinput_mice.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+        vertical = 1
+
+    error_context.context("Moving pointer from %s to %s" % (start_pos, end_pos))
+    console.pointer_move(end_pos, motion=line, mtype=mtype)
+    time.sleep(wait_time)
+
+    error_context.context("Collecting all pointer move events from guest")
+    while not events_queue.empty():
+        event = events_queue.get()
+        time.sleep(wait_time)
```
[the remark this answers is not in the document]
Comment (peixiu) [reply qemu/tests/vioinput_mice.py]: updated, thanks~

File: qemu/tests/vioinput_mice.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (peixiu) [reply qemu/tests/vioinput_mice.py]: I tested on all windows os(win7, win8, win8.1, win10, win2016, win2019), all can be passed with tolerance=40, so I keep this code as before.

**Comment by peixiu** [comment]
Tested results with latest codes:
 (1/9) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1022.23 s)
 (2/9) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1208.71 s)
 (3/9) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (3719.96 s)
 (4/9) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (162.77 s)
 (5/9) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (133.32 s)
 (6/9) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (78.16 s)
 (7/9) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (46.60 s)
 (8/9) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (71.32 s)
 (9/9) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (44.75 s)

 (1/4) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win7.i386.sp1.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (48.38 s)
 (2/4) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win7.i386.sp1.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (39.22 s)
 (3/4) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win7.x86_64.sp1.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (65.77 s)
 (4/4) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win7.x86_64.sp1.io-github-autotest-qemu.vioinput_mice.device_tablet.q35:PASS (55.42 s)

 (1/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (84.14 s)
 (2/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (83.55 s)

 (1/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.i386.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1027.67 s)
 (2/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.i386.1.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1019.75 s)
 (3/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.i386.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (76.81 s)
 (4/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.i386.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (53.84 s)
 (5/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.i386.1.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (63.95 s)
 (6/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.i386.1.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (54.05 s)

 (1/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.x86_64.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1298.97 s)
 (2/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.x86_64.1.io-github-autotest-qemu.unattended_install.cdrom.extra_cdrom_ks.default_install.aio_threads.q35: PASS (1354.52 s)
 (3/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.x86_64.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (85.18 s)
 (4/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.x86_64.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (54.83 s)
 (5/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.x86_64.1.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (76.63 s)
 (6/6) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win8.x86_64.1.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (52.20 s)

 (1/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (52.19 s)
 (2/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (51.62 s)

 (1/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (103.23 s)
 (2/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (46.83 s)

**Comment by peixiu** [comment]
@luckyh Please help to review it again when you free~
Thanks a lot~

**Comment by peixiu** [comment]
@luckyh when you review this patch, please help to review these 2 patch as well, they just adjust cfg files, both based on this patch,  thanks~ https://github.com/autotest/tp-qemu/pull/2042, https://github.com/autotest/tp-qemu/pull/2043

**Comment by peixiu** [comment]
@luckyh please help to review this patch again, thanks a lot~

File: qemu/tests/vioinput_mice.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+    if (xn - x0) != 0:
+        vertical = 0
+        k = (yn - y0) / (xn - x0)
+        b = yn - (k * xn)
+    else:
+        vertical = 1
+
+    error_context.context("Moving pointer from %s to %s" % (start_pos, end_pos))
+    console.pointer_move(end_pos, motion=line, _abs=_abs)
+    time.sleep(move_duration + wait_time)
```
Comment (luckyh) [inline qemu/tests/vioinput_mice.py]: Just realize that `time.sleep(wait_time)` is ok.

File: qemu/tests/vioinput_mice.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+
+    error_context.context("Collecting all pointer move events from guest")
+    while not events_queue.empty():
+        event = events_queue.get()
+        xpos, ypos = event["xPos"], event["yPos"]
+        # Filter beyond screen size events.
+        # Due to os will ignores/corrects these events.
```
Comment (luckyh) [inline qemu/tests/vioinput_mice.py]: Is there any official document pointed out this? if there is we can leave the link as a reference at here.

Comment (peixiu) [reply qemu/tests/vioinput_mice.py]: hi ,xuhan, that's sorry, we have not the doc for that, I confirmed before with vioinput developer through mail.

File: qemu/tests/vioinput_mice.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+        time.sleep(wait_time)
+        error_context.context("Check correct scroll event is received",
+                              logging.info)
+        while not events_queue.empty():
+            for i in range(count):
+                event = events_queue.get()
+                # some windows os will return pointer move event first
+                # before return scroll event, so filter them here.
+                if event["type"] == "POINTERMOVE":
+                    continue
```
[the remark this answers is not in the document]
Comment (peixiu) [reply qemu/tests/vioinput_mice.py]: Updated the code, thanks a lot~

File: qemu/tests/vioinput_mice.py (hunk 7 above)
Comment (peixiu) [reply qemu/tests/vioinput_mice.py]: Updated the code, thanks~

**Comment by peixiu** [comment]
Tested with latest code:
 (1/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.vioinput_mice.device_mouse.q35: PASS (46.26 s)
 (2/2) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.Win2012.x86_64.r2.io-github-autotest-qemu.vioinput_mice.device_tablet.q35: PASS (45.17 s)

@luckyh please help to review again~ thanks

**Comment by luckyh** [comment]
Thanks all.

**Pull request merged**